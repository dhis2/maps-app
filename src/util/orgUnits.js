import { uniqBy } from 'lodash/fp';
import i18n from '@dhis2/d2-i18n';
import { apiFetch } from './api';
import { getDisplayPropertyUrl } from './helpers';
import { getUniqueColor } from './colors';
import { qualitativeColors } from '../constants/colors';
import {
    ORG_UNIT_COLOR,
    ORG_UNIT_RADIUS,
    ORG_UNIT_RADIUS_SMALL,
    STYLE_TYPE_COLOR,
    STYLE_TYPE_SYMBOL,
    NONE,
} from '../constants/layers';

const getGroupColor = groups => {
    const groupsWithoutColors = groups.filter(g => !g.color);
    const colors = getUniqueColor(qualitativeColors);
    return group => {
        const index = groupsWithoutColors.findIndex(g => g.id === group.id);
        return colors(index);
    };
};

// Default symbol 21-25 are coloured circles
const getGroupSymbol = groups => {
    const groupsWithoutSymbols = groups.filter(g => !g.symbol);
    return group => {
        const index = groupsWithoutSymbols.findIndex(g => g.id === group.id);
        return index < 5 ? `${21 + index}.png` : '25.png';
    };
};

const parseGroupSet = ({ organisationUnitGroups: groups }) => {
    groups.sort((a, b) => a.name.localeCompare(b.name));

    const getColor = getGroupColor(groups);
    const getSymbol = getGroupSymbol(groups);

    return groups.map(group => ({
        ...group,
        color: group.color || getColor(group),
        symbol: group.symbol || getSymbol(group),
    }));
};

export const fetchOrgUnitGroupSet = id =>
    apiFetch(
        `/organisationUnitGroupSets/${id}?fields=organisationUnitGroups[id,name,color,symbol]`
    ).then(parseGroupSet);

export const fetchOrgUnitFields = () =>
    apiFetch(
        `/attributes.json?fields=id,name&filter=valueType:eq:GEOJSON&filter=organisationUnitAttribute:eq:true`
    ).then(({ attributes }) => attributes);

export const filterPointFacilities = data => data.filter(d => d.ty === 1);

export const getOrgUnitStyle = (dimensions, groupSet) =>
    groupSet &&
    groupSet.organisationUnitGroups &&
    dimensions &&
    dimensions[groupSet.id]
        ? groupSet.organisationUnitGroups.find(
              g => g.id === dimensions[groupSet.id]
          )
        : {};

export const getOrgUnitGroupLegendItems = (
    groups = [],
    useColor,
    contextPath
) =>
    groups.map(({ name, color = true, symbol }) =>
        useColor
            ? {
                  name,
                  color,
              }
            : {
                  name,
                  image: `${contextPath}/images/orgunitgroup/${symbol}`,
              }
    );

export const getStyledOrgUnits = (
    features = [],
    groupSet = {},
    { organisationUnitColor = ORG_UNIT_COLOR, radiusLow = ORG_UNIT_RADIUS },
    contextPath,
    orgUnitLevels
) => {
    const {
        name,
        styleType = orgUnitLevels ? STYLE_TYPE_COLOR : STYLE_TYPE_SYMBOL,
        organisationUnitGroups = [],
    } = groupSet;
    const isFacilityLayer = !orgUnitLevels;
    let levelWeight;
    let levelItems = [];

    if (!isFacilityLayer) {
        const levels = uniqBy(f => f.properties.level, features)
            .map(f => f.properties.level)
            .sort();

        levelWeight = level =>
            Math.pow(levels.length - levels.indexOf(level), 1.2);

        levelItems = levels.map(level => ({
            name: orgUnitLevels[level],
            color: organisationUnitColor,
            weight: levelWeight(level),
        }));
    }

    const useColor = styleType === STYLE_TYPE_COLOR;

    let styledFeatures = features.map(f => {
        const isPoint = f.geometry.type === 'Point';
        const { hasAdditionalGeometry } = f.properties;
        const { color, symbol } = getOrgUnitStyle(
            f.properties.dimensions,
            groupSet
        );
        let radius;

        if (isPoint) {
            radius = hasAdditionalGeometry
                ? ORG_UNIT_RADIUS_SMALL + 1
                : radiusLow;
        }

        const properties = {
            ...f.properties,
            radius,
        };

        if (useColor && color) {
            properties.color = hasAdditionalGeometry ? ORG_UNIT_COLOR : color;
        } else if (symbol) {
            properties.iconUrl = `${contextPath}/images/orgunitgroup/${symbol}`;
        }

        if (properties.level && levelWeight) {
            properties.weight = levelWeight(f.properties.level);
        }

        return {
            ...f,
            properties,
        };
    });

    // Only include facilities having a group membership
    if (isFacilityLayer && groupSet.id && !useColor) {
        styledFeatures = styledFeatures.filter(f => f.properties.iconUrl);
    }

    const groupItems = getOrgUnitGroupLegendItems(
        organisationUnitGroups,
        useColor,
        contextPath
    );

    const facilityItems =
        isFacilityLayer && !groupSet.id
            ? [
                  {
                      name: i18n.t('Facility'),
                      color: organisationUnitColor,
                      radius: radiusLow,
                  },
              ]
            : [];

    return {
        styledFeatures,
        legend: {
            unit: name,
            items: [...levelItems, ...groupItems, ...facilityItems],
        },
    };
};

// This function returns the org unit level names used in the legend
export const getOrgUnitLevels = async d2 => {
    const orgUnitLevels = await d2.models.organisationUnitLevels.list({
        fields: `id,${getDisplayPropertyUrl(d2)},level`,
        paging: false,
    });

    return orgUnitLevels
        ? orgUnitLevels.toArray().reduce(
              (obj, item) => ({
                  ...obj,
                  [item.level]: item.name,
              }),
              {}
          )
        : {};
};

// Loads default org unit level and group set for facility layer
export const fetchFacilityConfigurations = async () => {
    const [facilityOrgUnitLevel, facilityOrgUnitGroupSet] = await Promise.all([
        apiFetch('/configuration/facilityOrgUnitLevel'),
        apiFetch('/configuration/facilityOrgUnitGroupSet'),
    ]);

    return {
        facilityOrgUnitLevel,
        facilityOrgUnitGroupSet,
    };
};

// Returns coordinate field from layer config
export const getCoordinateField = ({ orgUnitField, orgUnitFieldDisplayName }) =>
    orgUnitField && orgUnitField !== NONE
        ? { id: orgUnitField, name: orgUnitFieldDisplayName }
        : null;

// Set hasAddiditionalGeometry property if exist
export const setAdditionalGeometry = features =>
    features
        .filter(
            (feature, i) =>
                i <
                features.findIndex(f => f.id === feature.id && f !== feature)
        )
        .forEach(f => (f.properties.hasAdditionalGeometry = true));
