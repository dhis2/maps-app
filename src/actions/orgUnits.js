import log from 'loglevel';
import * as types from '../constants/actionTypes';
import { fetchOrgUnits } from '../util/requests';

// Load organisation unit tree
export const loadOrgUnitTree = () => ({
    type: types.ORGANISATION_UNIT_TREE_LOAD,
});

// set organisation unit tree
export const setOrgUnitTree = rootModel => ({
    type: types.ORGANISATION_UNIT_TREE_SET,
    payload: rootModel,
});

// Load all organisation unit groups sets
export const loadOrgUnitLevels = () => ({
    type: types.ORGANISATION_UNIT_LEVELS_LOAD,
});

// Set all organisation unit groups sets
export const setOrgUnitLevels = data => ({
    type: types.ORGANISATION_UNIT_LEVELS_SET,
    payload: data,
});

// Load all organisation unit groups sets
export const loadOrgUnitGroups = () => ({
    type: types.ORGANISATION_UNIT_GROUPS_LOAD,
});

// Set all organisation unit groups sets
export const setOrgUnitGroups = data => ({
    type: types.ORGANISATION_UNIT_GROUPS_SET,
    payload: data,
});

// Load all organisation unit groups sets
export const loadOrgUnitGroupSets = () => ({
    type: types.ORGANISATION_UNIT_GROUP_SETS_LOAD,
});

// Set all organisation unit groups sets
export const setOrgUnitGroupSets = data => ({
    type: types.ORGANISATION_UNIT_GROUP_SETS_SET,
    payload: data,
});

export const setOrgUnitProfile = id => ({
    type: types.ORGANISATION_UNIT_PROFILE_SET,
    payload: id,
});

export const closeOrgUnitProfile = () => ({
    type: types.ORGANISATION_UNIT_PROFILE_CLOSE,
});

export const tSetOrgUnitTree = () => async dispatch => {
    try {
        const orgUnitTree = await fetchOrgUnits();
        dispatch(setOrgUnitTree(orgUnitTree));
    } catch (e) {
        log.error('Could not load organisation unit tree');
        return e;
    }
};
