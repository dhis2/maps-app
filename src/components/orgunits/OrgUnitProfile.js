import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import i18n from '@dhis2/d2-i18n';
import { IconCross24 } from '@dhis2/ui';
import Drawer from '../core/Drawer';
import OrgUnitInfo from './OrgUnitInfo';
import OrgUnitData from './OrgUnitData';
import { apiFetch } from '../../util/api';
import { getFixedPeriodsByType, filterFuturePeriods } from '../../util/periods';
import { closeOrgUnitProfile } from '../../actions/orgUnits';
import styles from './styles/OrgUnitProfile.module.css';

// Only YEARLY period type is supported in first version
const periodType = 'YEARLY';
const currentYear = String(new Date().getFullYear());
const periods = getFixedPeriodsByType(periodType, currentYear);
const defaultPeriod = filterFuturePeriods(periods)[0] || periods[0];

/*
 *  Loads an org unit profile and displays it in a right drawer component
 */
export const OrgUnitProfile = ({ id, closeOrgUnitProfile }) => {
    const [profile, setProfile] = useState();

    // Load org unit profile when id is changed
    // https://docs.dhis2.org/en/develop/using-the-api/dhis-core-version-master/org-unit-profile.html
    useEffect(() => {
        if (id) {
            apiFetch(
                `/organisationUnitProfile/${id}/data?period=${defaultPeriod.id}`
            ).then(setProfile);
        }
    }, [id]);

    if (!id) {
        return null;
    }

    return (
        <Drawer className={styles.drawer}>
            <div className={styles.header}>
                {i18n.t('Organisation unit profile')}
                <span className={styles.close} onClick={closeOrgUnitProfile}>
                    <IconCross24 />
                </span>
            </div>
            <div className={styles.content}>
                {profile && (
                    <>
                        <OrgUnitInfo
                            {...profile.info}
                            groupSets={profile.groupSets}
                            attributes={profile.attributes}
                        />
                        <OrgUnitData
                            id={id}
                            periodType={periodType}
                            defaultPeriod={defaultPeriod}
                            data={profile.dataItems}
                        />
                    </>
                )}
            </div>
        </Drawer>
    );
};

OrgUnitProfile.propTypes = {
    id: PropTypes.string,
    closeOrgUnitProfile: PropTypes.func.isRequired,
};

export default connect(
    ({ orgUnitProfile }) => ({
        id: orgUnitProfile,
    }),
    { closeOrgUnitProfile }
)(OrgUnitProfile);
