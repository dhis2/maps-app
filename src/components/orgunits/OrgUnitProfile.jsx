import { useDataQuery } from '@dhis2/app-runtime'
import i18n from '@dhis2/d2-i18n'
import { CenteredContent, CircularLoader, IconCross24 } from '@dhis2/ui'
import React, { useEffect } from 'react'
import { useSelector, useDispatch } from 'react-redux'
import { closeOrgUnitProfile } from '../../actions/orgUnits.js'
import useKeyDown from '../../hooks/useKeyDown.js'
import Drawer from '../core/Drawer.jsx'
import OrgUnitData from './OrgUnitData.jsx'
import OrgUnitInfo from './OrgUnitInfo.jsx'
import styles from './styles/OrgUnitProfile.module.css'

const ORGUNIT_PROFILE_QUERY = {
    profile: {
        resource: 'organisationUnitProfile',
        id: ({ id }) => `${id}/data`,
    },
}

/*
 *  Loads an org unit profile and displays it in a right drawer component
 */
const OrgUnitProfile = () => {
    const id = useSelector((state) => state.orgUnitProfile)
    const dispatch = useDispatch()
    const { loading, data, refetch } = useDataQuery(ORGUNIT_PROFILE_QUERY, {
        lazy: true,
    })

    useEffect(() => {
        if (id) {
            refetch({
                id,
            })
        }
    }, [id, refetch])

    useKeyDown('Escape', () => dispatch(closeOrgUnitProfile()), true)

    if (!id) {
        return null
    }

    return (
        <Drawer className={styles.drawer} dataTest="org-unit-profile">
            <div className={styles.header}>
                {i18n.t('Organisation unit profile')}
                <span
                    role="button"
                    aria-label={i18n.t('Close')}
                    className={styles.close}
                    onClick={() => dispatch(closeOrgUnitProfile())}
                >
                    <IconCross24 />
                </span>
            </div>
            <div className={styles.content}>
                {loading && (
                    <CenteredContent>
                        <CircularLoader />
                    </CenteredContent>
                )}
                {!loading && data?.profile && (
                    <>
                        <OrgUnitInfo
                            {...data.profile.info}
                            groupSets={data.profile.groupSets}
                            attributes={data.profile.attributes}
                        />
                        <OrgUnitData id={id} />
                    </>
                )}
            </div>
        </Drawer>
    )
}

export default OrgUnitProfile
