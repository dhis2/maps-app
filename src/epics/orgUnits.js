import { combineEpics } from 'redux-observable';
import { getInstance as getD2 } from 'd2/lib/d2';
import 'rxjs/add/operator/concatMap';
import * as types from '../constants/actionTypes';
import { errorActionCreator } from '../actions/helpers';
import { getDisplayPropertyUrl } from '../util/helpers';
import { apiFetch } from '../util/api';
import {
    setOrgUnit,
    setOrgUnitTree,
    setOrgUnitLevels,
    setOrgUnitGroups,
    setOrgUnitGroupSets,
    setOrgUnitCoordinate,
} from '../actions/orgUnits';
import { setOrgUnitPath } from '../actions/layerEdit';

export const loadOrgUnit = action$ =>
    action$.ofType(types.ORGANISATION_UNIT_LOAD).concatMap(action =>
        getD2()
            .then(d2 =>
                d2.models.organisationUnit.get(action.payload.id, {
                    fields: `id,${getDisplayPropertyUrl(
                        d2
                    )},code,address,email,phoneNumber,coordinates,parent[id,${getDisplayPropertyUrl(
                        d2
                    )}],organisationUnitGroups[id,${getDisplayPropertyUrl(
                        d2
                    )}]`,
                })
            )
            .then(setOrgUnit)
    );

export const loadOrgUnitTree = action$ =>
    action$.ofType(types.ORGANISATION_UNIT_TREE_LOAD).concatMap(() =>
        getD2()
            .then(d2 =>
                d2.models.organisationUnits.list({
                    level: 1,
                    fields:
                        'id,path,displayName,children[id,path,displayName,children::isNotEmpty]',
                })
            )
            .then(modelCollection => setOrgUnitTree(modelCollection.toArray()))
            .catch(errorActionCreator(types.ORGANISATION_UNIT_TREE_LOAD_ERROR))
    );

export const loadOrgUnitLevels = action$ =>
    action$.ofType(types.ORGANISATION_UNIT_LEVELS_LOAD).concatMap(() =>
        getD2()
            .then(async d2 =>
                d2.models.organisationUnitLevels.list({
                    fields: `id,${getDisplayPropertyUrl(d2)},level`,
                    pageing: false,
                })
            )
            .then(levels => setOrgUnitLevels(levels.toArray()))
            .catch(
                errorActionCreator(types.ORGANISATION_UNIT_LEVELS_LOAD_ERROR)
            )
    );

export const loadOrgUnitGroups = action$ =>
    action$.ofType(types.ORGANISATION_UNIT_GROUPS_LOAD).concatMap(() =>
        getD2()
            .then(async d2 =>
                d2.models.organisationUnitGroups.list({
                    fields: `id,${getDisplayPropertyUrl(d2)}`,
                    pageing: false,
                })
            )
            .then(groups => setOrgUnitGroups(groups.toArray()))
            .catch(
                errorActionCreator(types.ORGANISATION_UNIT_GROUPS_LOAD_ERROR)
            )
    );

export const loadOrgUnitGroupSets = action$ =>
    action$.ofType(types.ORGANISATION_UNIT_GROUP_SETS_LOAD).concatMap(() =>
        getD2()
            .then(d2 =>
                d2.models.organisationUnitGroupSets.list({
                    fields: `id,${getDisplayPropertyUrl(d2)}`,
                    pageing: false,
                })
            )
            .then(groupSets => setOrgUnitGroupSets(groupSets.toArray()))
            .catch(
                errorActionCreator(
                    types.ORGANISATION_UNIT_GROUP_SETS_LOAD_ERROR
                )
            )
    );

export const changeOrgUnitCoordinate = action$ =>
    action$
        .ofType(types.ORGANISATION_UNIT_COORDINATE_CHANGE)
        .concatMap(({ layerId, featureId, coordinate }) =>
            apiFetch(`/organisationUnits/${featureId}`, 'PATCH', {
                coordinates: JSON.stringify(coordinate),
            }).then(response => {
                if (response.ok) {
                    return setOrgUnitCoordinate(layerId, featureId, coordinate); // Update org. unit in redux store
                } else {
                    return errorActionCreator(
                        types.ORGANISATION_UNIT_COORDINATE_CHANGE_ERROR
                    );
                }
            })
        );

// Load organisation unit tree path (temporary solution, as favorites don't include paths)
export const loadOrgUnitPath = action$ =>
    action$
        .ofType(types.LAYER_EDIT_ORGANISATION_UNIT_PATH_LOAD)
        .concatMap(action =>
            getD2()
                .then(async d2 =>
                    d2.models.organisationUnit.get(action.id, {
                        fields: 'path',
                    })
                )
                .then(ou => setOrgUnitPath(action.id, ou.path))
                .catch(
                    errorActionCreator(
                        types.LAYER_EDIT_ORGANISATION_UNIT_PATH_LOAD_ERROR
                    )
                )
        );

export default combineEpics(
    loadOrgUnit,
    loadOrgUnitTree,
    loadOrgUnitLevels,
    loadOrgUnitGroups,
    loadOrgUnitGroupSets,
    changeOrgUnitCoordinate,
    loadOrgUnitPath
);
