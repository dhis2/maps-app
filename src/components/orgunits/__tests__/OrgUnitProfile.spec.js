import { render, fireEvent, screen } from '@testing-library/react'
import React from 'react'
import { act } from 'react-dom/test-utils.js'
import { Provider } from 'react-redux'
import configureMockStore from 'redux-mock-store'
import { closeOrgUnitProfile } from '../../../actions/orgUnits.js'
import OrgUnitProfile from '../OrgUnitProfile.js'

jest.mock('../../../util/api.js', () => {
    return {
        apiFetch: jest.fn().mockResolvedValue({}),
    }
})

jest.mock('../../../actions/orgUnits', () => {
    return {
        closeOrgUnitProfile: jest
            .fn()
            .mockReturnValue({ type: 'ORGANISATION_UNIT_PROFILE_CLOSE' }),
    }
})

jest.mock(
    '../OrgUnitData',
    () =>
        function MockOrgUnitData() {
            return <div>OrgUnitData</div>
        }
)

jest.mock(
    '../OrgUnitInfo',
    () =>
        function MockOrgUnitInfo() {
            return <div>OrgUnitInfo</div>
        }
)

const mockStore = configureMockStore()

describe('Org unit profile (location details)', () => {
    test('does not render if no org unit profile is set', () => {
        const store = {
            orgUnitProfile: null,
        }

        const { container } = render(
            <Provider store={mockStore(store)}>
                <OrgUnitProfile />
            </Provider>
        )
        expect(container).toMatchSnapshot()
    })

    test('renders if org unit profile is set', async () => {
        const promise = Promise.resolve()
        const store = {
            orgUnitProfile: 'rainbowDash',
        }

        const { container } = render(
            <Provider store={mockStore(store)}>
                <OrgUnitProfile />
            </Provider>
        )
        await act(() => promise)
        expect(container).toMatchSnapshot()
    })

    // TODO: this is testing implementation details. Instead it should
    // test that the component renders null after the button is clicked
    test('calls closeOrgUnitProfile when X button clicked', async () => {
        const promise = Promise.resolve()
        const store = {
            orgUnitProfile: 'rainbowDash',
        }

        const { container } = render(
            <Provider store={mockStore(store)}>
                <OrgUnitProfile />
            </Provider>
        )

        await act(() => promise)
        expect(container).toMatchSnapshot()

        fireEvent.click(screen.getByLabelText('Close'))

        expect(closeOrgUnitProfile).toHaveBeenCalled()
    })
})
