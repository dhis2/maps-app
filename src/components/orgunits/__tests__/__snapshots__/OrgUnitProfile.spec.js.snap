// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Org unit profile (location details) calls closeOrgUnitProfile when X button clicked 1`] = `
<div>
  <div
    class="drawer drawer right"
  >
    <div
      class="header"
    >
      Organisation unit profile
      <span
        aria-label="Close"
        class="close"
        role="button"
      >
        <svg
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M6.613 5.21l.094.083L12 10.585l5.293-5.292a1 1 0 011.497 1.32l-.083.094L13.415 12l5.292 5.293a1 1 0 01-1.32 1.497l-.094-.083L12 13.415l-5.293 5.292a1 1 0 01-1.497-1.32l.083-.094L10.585 12 5.293 6.707a1 1 0 011.32-1.497z"
            fill="currentColor"
          />
        </svg>
      </span>
    </div>
    <div
      class="content"
    >
      <div>
        OrgUnitInfo
      </div>
      <div>
        OrgUnitData
      </div>
    </div>
  </div>
</div>
`;

exports[`Org unit profile (location details) does not render if no org unit profile is set 1`] = `<div />`;

exports[`Org unit profile (location details) renders if org unit profile is set 1`] = `
<div>
  <div
    class="drawer drawer right"
  >
    <div
      class="header"
    >
      Organisation unit profile
      <span
        aria-label="Close"
        class="close"
        role="button"
      >
        <svg
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M6.613 5.21l.094.083L12 10.585l5.293-5.292a1 1 0 011.497 1.32l-.083.094L13.415 12l5.292 5.293a1 1 0 01-1.32 1.497l-.094-.083L12 13.415l-5.293 5.292a1 1 0 01-1.497-1.32l.083-.094L10.585 12 5.293 6.707a1 1 0 011.32-1.497z"
            fill="currentColor"
          />
        </svg>
      </span>
    </div>
    <div
      class="content"
    >
      <div>
        OrgUnitInfo
      </div>
      <div>
        OrgUnitData
      </div>
    </div>
  </div>
</div>
`;
